api/repositories/app_repository_test.go

```diff
@@ -28,6 +31,263 @@ const (
 	CFAppStoppedState  = "STOPPED"
 )
 
+//counterfeiter:generate -o fake -fake-name ClientWithWatch sigs.k8s.io/controller-runtime/pkg/client.WithWatch
+var _ = Describe("AppRepository Units", func() {
```

Comment:
I guess we'd split the integration tests out into a separate dir when this goes out of draft?

Reply:
We were on the fence about that. We are currently in a strange state because we've been using integration level tests as unit tests, and measuring coverage against those tests. This PR backfills some testing to get to unreachable code, but doesn't attempt to get to full coverage on the unit test side.

Ideally, we would stop using integration tests as unit tests, and write unit tests that achieve full coverage on their own, and move the env tests into a separate suite, but that would take a lot of time, so in the interim we settled on putting both in the same file to make it easier to understand how they relate to one another to achieve test coverage.

So, yeah basically, I agree that the two kinds of tests should be separated, but I am less sure that immediately is the right time frame, unless we were going to tear off the bandage and rewrite all the tests.

Reply:
That's something we wanted feedback on. In the past we've put unit and integration tests in separate packages. This makes it harder to find all of the tests and requires you to remember the structure of each different package. Having the unit tests in the same file with the integration tests solves this problem, but at the cost of slightly higher overhead per test (due to the BeforeEach targeted at test-env tests)

Reply:
Even if we do not convert integrations to units right away I still think it is good to have a separation between those two. If they are separated we can agree on some convention of what goes where. For example we might agree that we only ever put error cases in the units. But if there are no integrations and units it is hard to even communicate that and I am afraid these suites will continue growing in a diffuse manner.

Reply:
I hear that some of the reasoning behind mixing the pure units with the test-env integrations might be that this way it is easier to see combined code coverage in the IDE. However not everyone uses the same IDE/editor, while we do have code climate [coverage](https://codeclimate.com/github/cloudfoundry/cf-k8s-controllers) setup that is merging coverage from all tests no matter if they are in the same package or not (with the exception of e2es). To me this is a better place to look for uncovered branches as it is central and we can create chores with links to the reports.